assets/app/scripts/directives/buildTrendsChart.js

```diff
@@ -0,0 +1,252 @@
+        };
+
+        var numCompleteBuilds = function() {
+          updateCompleteBuilds();
+          return $scope.completeBuilds.length;
+        };
+
+        var annotationFilter = $filter('annotation');
+        var getBuildNumber = function(build) {
+          var buildNumber = annotationFilter(build, 'buildNumber') || parseInt(build.metadata.name.match(/(\d+)$/), 10);
+          if (isNaN(buildNumber)) {
+            return null;
+          }
+
+          return buildNumber;
+        };
+
+        var getDuration = function(build) {
+          var startTimestamp = getStartTimestsamp(build);
+          var endTimestamp = build.status.completionTimestamp;
+          if (!startTimestamp || !endTimestamp) {
+            return 0;
+          }
+
+          return moment(endTimestamp).diff(moment(startTimestamp));
+        };
+
+        var chart, averageDuration, showAverageLine = false;
+        var updateAvgLine = function() {
+          if (showAverageLine) {
+            chart.ygrids([{
+              value: averageDuration,
+              'class': 'build-trends-avg-line'
+            }]);
+          } else {
+            chart.ygrids.remove();
+          }
+        };
+
+        $scope.toggleAvgLine = function() {
+          showAverageLine = !showAverageLine;
+          updateAvgLine();
+        };
+
+        var update = function() {
+          // Keep a map of builds by number so we can find the build later when a data point is clicked.
+          buildByNumber = {};
+          var data = {
+            json: [],
+            keys: {
+              x: 'buildNumber'
+            }
+          };
+
+          var sum = 0, count = 0, foundPhases = {};
+          angular.forEach($scope.completeBuilds, function(build) {
+            var buildNumber = getBuildNumber(build);
+            if (!buildNumber) {
+              return;
+            }
+
+            var duration = getDuration(build);
+
+            // Track the sum and count to calculate the average duration.
+            sum += duration;
+            count++;
+
+            var buildData = {
+              buildNumber: buildNumber
+            };
+            buildData[build.status.phase] = duration;
+            data.json.push(buildData);
+            buildByNumber[buildNumber] = build;
+            foundPhases[build.status.phase] = true;
+          });
+
+          // Show only the last 50 builds.
+          if (data.json.length > 50) {
+            data.json.sort(function(a, b) {
+              return a.buildNumber - b.buildNumber;
+            });
+            data.json = data.json.slice(data.json.length - 50);
+          }
+
+          averageDuration = sum / count;
+          $scope.averageDurationText = humanize(averageDuration);
+
+          // Only show found phases in the legend.
+          var groups = _.filter(completePhases, function(phase) {
+            return foundPhases[phase];
+          });
+          data.keys.value = groups;
+          data.groups = [groups];
```

Comment:
same thing, make sure this doesnt break if none of the completed builds had build numbers